### DISCUSSION TITLE
'AuditLogDiff' object has no attribute 'type'

### DISCUSSION THREAD

**Discussion by ghost** [Rapptz/discord.py#5842] [Q&A] [answered]
## Summary

AuditLogDiff seems to be missing the 'type' attribute despite the official API docs clearly saying that AuditLogDiff *should* have an attribute 'type'. 

## Reproduction Steps

I simply ran the following code
```
@client.event
async def on_guild_channel_update(channelold,channelnew):
    print('channel "'+str(channelold.name)+'" (now called "'+str(channelnew.name)+'" was updated!')
    # Grab audit log
    updatelog = await channelnew.guild.audit_logs(limit=1,action=discord.AuditLogAction.channel_update).flatten()
    print(str(updatelog[0].user.name))
    cName = False
    cPos = False
    cNSFW = False
    cPerms = False
    cTopic = False
    cSlow = False
    
    cBefore = updatelog[0].before
    cAfter = updatelog[0].after
    
    if cBefore.type == 1:
        return
```

## Expected Results

The code snippet would run perfectly fine without errors, continuing if the channel was a text channel, and stopping if it was a voice channel.

## Actual Results

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/discord/client.py", line 312, in _run_event
    await coro(*args, **kwargs)
  File "./etp_bot/bot.py", line 131, in on_guild_channel_update
    if cBefore.type == 1:
AttributeError: 'AuditLogDiff' object has no attribute 'type'
```

## Checklist

- [x] I have searched the open issues for duplicates.
- [x] I have shown the entire traceback, if possible.
- [x] I have removed my token from display, if visible.

## System Information

- Python v3.7.3-final
- discord.py v1.4.1-final
- aiohttp v3.6.2
- system info: Linux 4.19.113+ #1300 Thu Mar 26 16:47:42 GMT 2020

## Notes

(by the way, I am aware I could easily use `channelold.type` but a bug is still a bug and that workaround may not always be available)

**Comment by Vexs** [comment]
These attrs are added at runtime from the API payloads- if the API doesn't provide a key, it won't be included in the object, which is documented. 
Given the shifting nature of discord's API I wouldn't necessarily rely on any fields being there frankly.

**Comment by SebbyLaw** [comment, marked as answer]
~~According to the [discord documentation](https://discord.com/developers/docs/resources/audit-log#audit-log-change-object-audit-log-change-key), `type` is implied to be used only when a channel is created:~~
~~> type of entity created~~

~~> (by the way, I am aware I could easily use `channelold.type` but a bug is still a bug and that workaround may not always be available)~~

~~`This is false.~~

Edit: Apologies for the rudeness and the semi-wrong answer. `type` will only be a valid property when the channel type is changed, which currently only happens when a text channel becomes a news channel.

**Comment by Rapptz** [comment]
The attribute `AuditLogDiff.type` depends on the `AuditLogAction` that is being iterated on. You can see which audit log actions support the `type `attribute in the [documentation](https://discordpy.readthedocs.io/en/latest/api.html#discord.AuditLogAction).
